anxcloud/resource_kubernetes_cluster_test.go

```diff
@@ -0,0 +1,172 @@
+					token                  = anxcloud_kubernetes_kubeconfig.foo.token
+					cluster_ca_certificate = base64decode(anxcloud_kubernetes_kubeconfig.foo.cluster_ca_certificate)
+				}
+				
+				resource "kubernetes_namespace" "foo" {
+					metadata {
+						name = "foo"
+					}
+				}
+
+				`, cluster.Name),
+				Check: resource.ComposeAggregateTestCheckFunc(
+					testAccAnxCloudKubernetesClusterExists(&cluster, &nodePool1),
+					resource.TestCheckResourceAttrPair("anxcloud_kubernetes_cluster.foo", "id", "data.anxcloud_kubernetes_cluster.foo", "id"),
+					resource.TestCheckResourceAttrPair("anxcloud_kubernetes_cluster.foo", "location", "data.anxcloud_kubernetes_cluster.foo", "location"),
+					resource.TestCheckResourceAttrPair("anxcloud_kubernetes_cluster.foo", "node_pool", "data.anxcloud_kubernetes_cluster.foo", "node_pool"),
+				),
+			},
+		},
+	})
+}
+
+func testAccAnxCloudKubernetesCluster(cluster *kubernetesv1.Cluster, nodePools ...*kubernetesv1.NodePool) string {
+	return fmt.Sprintf(`
+	resource "anxcloud_kubernetes_cluster" "foo" {
+		name = "%s"
+		location = "%s"
+		deploy_service_vms = %v
+
+		%s # node pools
+	}
+	`,
+		cluster.Name,
+		cluster.Location.Identifier,
+		*cluster.NeedsServiceVMs,
+		strings.Join(mapSlice(testAccAnxCloudKubernetesNodePool, nodePools), "\n"),
+	)
+}
+
+func testAccAnxCloudKubernetesNodePool(nodePool *kubernetesv1.NodePool) string {
+	return fmt.Sprintf(`
+	node_pool {
+		name = "%s"
+		replicas = %d
+		memory = 4
+		disk_size = 10
+		cpus = 2
+	}
+	`,
+		nodePool.Name,
+		*nodePool.Replicas,
+	)
+}
+
+func testAccAnxCloudKubernetesClusterExists(cluster *kubernetesv1.Cluster, nodePools ...*kubernetesv1.NodePool) resource.TestCheckFunc {
+	n := "anxcloud_kubernetes_cluster.foo"
+
+	return func(s *terraform.State) error {
+		rs, ok := s.RootModule().Resources[n]
+		if !ok {
+			return fmt.Errorf("resource not found in state: %s", n)
+		}
+
+		a := apiFromProviderConfig(testAccProvider.Meta())
+
+		engineCluster := kubernetesv1.Cluster{Identifier: rs.Primary.ID}
+		if err := a.Get(context.TODO(), &engineCluster); err != nil {
+			return fmt.Errorf("failed retrieving kubernetes cluster: %s", err)
+		}
+
+		expectedNodePoolNames := mapSlice(func(np *kubernetesv1.NodePool) string {
+			return np.Name
+		}, nodePools)
+
+		ctx, cancel := context.WithTimeout(context.TODO(), 20*time.Second)
+		defer cancel()
+
+		var channel types.ObjectChannel
+		if err := a.List(ctx, &kubernetesv1.NodePool{Cluster: engineCluster}, api.ObjectChannel(&channel)); err != nil {
+			return fmt.Errorf("failed listing node pools of cluster: %s", err)
+		}
+
+		nodePool := kubernetesv1.NodePool{}
+		for retriever := range channel {
+			if err := retriever(&nodePool); err != nil {
+				return fmt.Errorf("failed retrieving node pool: %s", err)
+			}
+
+			expectedNodePoolNames = filterSlice(func(name string) bool {
+				return name != nodePool.Name
+			}, expectedNodePoolNames)
+		}
+
+		if len(expectedNodePoolNames) != 0 {
```

Comment:
correctly checking for all expected node pools being here, but no check for "no additional node pool existing"

Reply:
Outdated